import libc.(abort);



//
// error (exceptions disabled)
//

errorNoThrow(...e) {
    printlnTo(stderr, "error: ", ...e);
    abort();
}

[ | not ExceptionsEnabled?]
error(...e) {
    errorNoThrow(...e);
}



//
// error (exceptions enabled)
//

record Error (msg : String);

instance Exception = Error;

[ | ExceptionsEnabled?]
overload error(...e) {
    var buf = String();
    printTo(buf, ...e);
    throw Error(move(buf));
}



//
// assert
//

callbyname assert(cond:Bool, ...message) {
    if (not cond)
        errorNoThrow("assertion failed: ", ...message);
}

callbyname overload assert(cond:Bool) {
    if (not cond)
        errorNoThrow("assertion failed");
}
